content/module2/6-traffic-routing.md

```diff
@@ -0,0 +1,112 @@
+
+
+```mermaid
+flowchart TB
+
+subgraph Reverse_Proxy[Reverse Proxy]
+    C2[Client] --> RP[Reverse Proxy]
+    RP --> S1[Backend Server 1]
+    RP --> S2[Backend Server 2]
+end
+
+C2:::client
+RP:::proxy
+
+classDef client fill:#cce5ff,stroke:#000,stroke-width:1px
+classDef proxy fill:#ffeeba,stroke:#000,stroke-width:1px
+
+```
+
+## Forward proxy
+This does the inverse where it hides the **client** rather than the server. The way that it works is that when entering a URL on a browser like `kalschatzi.com` it first connects to the configured proxy server and only then to the actual `kalschatzi.com` server, which can itself be exposed with a reverse proxy. This is a **client** side configuration and will not affect in any way the server
+
+### Why use one?
+There can be a number of benefits, depending on need.
+* **Caching**: proxies can cache GET requests and this speeds up responses and saves bandwidth
+* **Content Filtering / Control**: It allows content being request to be filtered and blocked if necessary.
+* **Privacy**: It can hide your real IP address.
+* **Security**: It can enforce security policies like only allow HTTPS sites
+
+```mermaid
+flowchart TB
+
+subgraph Forward_Proxy[Forward Proxy]
+    C1[Client] --> FP[Forward Proxy]
+    FP --> Internet[Public Internet]
+    Internet --> Site1[example.com]
+end
+
+C1:::client
+FP:::proxy
+
+classDef client fill:#cce5ff,stroke:#000,stroke-width:1px
+classDef proxy fill:#ffeeba,stroke:#000,stroke-width:1px
+
+```
+
+## Load Balancers
+
+Load balancers are similar to reverse proxies with a bigger focus on traffic shapping. Their main goal is ensure high availability and load distribution.
+It can handle things like if your application has multiple replicas/servers it will distribute the load across all of them, and if one of them is down or unresponsive, it will stop sending request to that one and only use the other ones.
+
+Some reverse proxies can also work as load balancers, like [NGINX](https://nginx.org/en/) (reads engine X).
+
+There are different kinds of load balancer, while the first 2 on the list are the most commonly used on day-to-day
+
+### Layer 4 Load Balancer
+
+Works at the TCP/UDP level. 
+Makes decisions based on:
+* IP address
+* Port number
+* Protocol (TCP/UDP)
+
+It's good for raw traffic routing, but if you want it to handle things like routing based on request URL (host header) that is a layer 7 feature and is not supported.
+
+### Layer 7 Load Balancer
+
+Works at the HTTP/HTTPS level
+Makes decisions based on:
+
+* URL path (/api, /images)
+* Host header (learn.kalschatzi.com)
+* Cookies, user-agent, other headers.
+
+This has the benefit of being able to do what layer 4 LB can do, modify requests like adding headers (`x-forwarded-from` for example) and do smarter routing based on Layer 7 properties.
+The downside is that it's heavier and slower that a TCP/UDP LB.
+
+### Global Load Balancer (DNS/Geo-Based)
+Routes users to the closest datacenter
+Based on:
```

Comment:
Pedantic comment - I think you can have this as a whole sentence - "Routes users to the closest datacenter based on:"